### PR TITLE
Fix otNetifAddress declaration to work between C & C++

### PR DESCRIPTION
**Pull request by nibanks**
The padding and field sizes (for bool) were different between C & C++. This gets them lined up.

### CODE REVIEW

**Comment by jwhui** [comment]
Just curious, what issue were you running into?

File: include/openthread-types.h [hunk 1]
```diff
@@ -895,16 +895,19 @@ typedef struct otBufferInfo
...
     uint32_t               mPreferredLifetime;       ///< The Preferred Lifetime.
     uint32_t               mValidLifetime;           ///< The Valid lifetime.
     uint8_t                mPrefixLength;            ///< The Prefix length.
-    unsigned int           mScopeOverride : 4;       ///< The IPv6 scope of this address.
-    bool                   mScopeOverrideValid : 1;  ///< TRUE if the mScopeOverride value is valid, FALSE othewrise.
+    uint32_t               mScopeOverride : 4;       ///< The IPv6 scope of this address.
+    uint32_t               mScopeOverrideValid : 1;  ///< TRUE if the mScopeOverride value is valid, FALSE othewrise.
     struct otNetifAddress *mNext;                    ///< A pointer to the next network interface address.
-} otNetifAddress;
+} OT_TOOL_PACKED_END;
```
Comment (pvanhorn) [inline include/openthread-types.h]: This packed macro is going to create un-optimized accessor instructions. Generally, if objects of a type are not crossing a medium between processors, such as a bus or radio, the packed macro should be avoided.  You should be able to achieve your goals without the packed macro anyways.

Comment (mszczodrak) [inline include/openthread-types.h]: :+1:

**Comment by nibanks** [comment]
There were a couple of issues here:

- `bool` sizes weren't necessarily the same between C & C++
- packing between the 4 byte and 1 byte fields wasn't necessarily the same
- packing at the end of the struct wasn't necessarily the same

**Comment by pvanhorn** [comment]
What compiler are you using?  I suspect there will be compiler options that you could employ to solve those issues. Likely, the options are different for the C compiler vs the C++ compiler.

**Comment by nibanks** [comment]
Microsoft Visual Studio

**Comment by pvanhorn** [comment]
I'm having a hard time believing your claims.  Can you provide output of sizeof(struct) and address of last struct element, from a C and C++ file for comparison sake to confirm your claims? If this is true it would create all sorts of problems that other programmers around the world would have encountered. I ha

**Comment by pvanhorn** [comment]
I did an online search and found reference to the /Zp option for MVS.  Perhaps you have this option set differently for C & C++?  Although from what I've read, setting this option should affect both C and C++ together.

**Comment by pvanhorn** [comment]
Alternatively perhaps you could use the /TP option which causes all files to be treated as C++ regardless of the file name suffix.

**Comment by nibanks** [comment]
I am experimenting a bit more, but it looks like all the issues came down to `bool` being typedef'd as an `int` in C. When I change that to a `_Bool` it fixes the problem.

**Comment by pvanhorn** [comment]
Excellent find.

File: tests/unit/test_toolchain.cpp [hunk 2]
```diff
@@ -93,12 +98,31 @@ void test_packed_enum()
     VerifyOrQuit(neighbor.mState == Thread::Neighbor::kStateValid, "Toolchain::OT_TOOL_PACKED failed 4\n");
 }
 
+void test_addr_size_cpp()
+{
+#ifdef X64
+    CompileTimeAssert(offsetof(otNetifAddress, mNext) == 40, "mNext should offset by 40 bytes from front");
+    CompileTimeAssert(sizeof(otNetifAddress) == 48, "otNetifAddress should be 48 (unpacked) bytes");
+#else
+    CompileTimeAssert(offsetof(otNetifAddress, mNext) == 36, "mNext should offset by 36 bytes from front");
```
Comment (pvanhorn) [inline tests/unit/test_toolchain.cpp]: These magic numbers create a bit of a maintenance headache.  If someone changes the data structure in a valid way they then may need to change these hard coded values.  Furthermore, they must know how to change the values to create a successful result. 

What I think you really want is a way to compare the result of sizeof(..) and offset(...) from the cpp file to the same result in the c file below.  For that you probably can't get away with using CompileTimeAssert but instead must do something like;

VerifyOrQuit(sizeof(otNetifAddress) == GetCSizeofOtNetifAddress())

where the function GetCSizeofNetifAddress is implemented in test_toolchain_c.c

Comment (nibanks) [inline tests/unit/test_toolchain.cpp]: Yeah, that makes sense. I can do that.

**Comment by nibanks** [comment]
Can you please give travis a nudge? Everything passed on my builds.

**Comment by pvanhorn** [comment]
LGTM

**Comment by pvanhorn** [comment]
done

**Pull request merged**